frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,670 @@
+		// be rebagged
+		for inserted_bag in new_bags {
+			let affected_bag = {
+				// this recreates `notional_bag_for` logic, but with the old thresholds.
+				let idx = old_thresholds.partition_point(|&threshold| inserted_bag > threshold);
+				old_thresholds.get(idx).copied().unwrap_or(VoteWeight::MAX)
+			};
+			if !affected_old_bags.insert(affected_bag) {
+				// If the previous threshold list was [10, 20], and we insert [3, 5], then there's
+				// no point iterating through bag 10 twice.
+				continue
+			}
+
+			if let Some(bag) = Bag::<T>::get(affected_bag) {
+				affected_accounts.extend(bag.iter().map(|node| node.id));
+			}
+		}
+
+		let removed_bags = old_set.difference(&new_set).copied();
+		// a removed bag means that all members of that bag must be rebagged
+		for removed_bag in removed_bags.clone() {
+			if !affected_old_bags.insert(removed_bag) {
+				continue
+			}
+
+			if let Some(bag) = Bag::<T>::get(removed_bag) {
+				affected_accounts.extend(bag.iter().map(|node| node.id));
+			}
+		}
+
+		// migrate the voters whose bag has changed
+		let weight_of = T::VoteWeightProvider::vote_weight;
+		Self::remove_many(&affected_accounts);
+		let num_affected = affected_accounts.len() as u32;
+		let _ = Self::insert_many(affected_accounts.into_iter(), weight_of);
+
+		// we couldn't previously remove the old bags because both insertion and removal assume that
+		// it's always safe to add a bag if it's not present. Now that that's sorted, we can get rid
+		// of them.
+		//
+		// it's pretty cheap to iterate this again, because both sets are in-memory and require no
+		// lookups.
+		for removed_bag in removed_bags {
+			debug_assert!(
+				!crate::ListNodes::<T>::iter().any(|(_, node)| node.bag_upper == removed_bag),
+				"no id should be present in a removed bag",
+			);
+			crate::ListBags::<T>::remove(removed_bag);
+		}
+
+		debug_assert!(
+			{
+				let thresholds = T::BagThresholds::get();
+				crate::ListBags::<T>::iter().all(|(threshold, _)| thresholds.contains(&threshold))
+			},
+			"all `bag_upper` in storage must be members of the new thresholds",
+		);
+
+		num_affected
+	}
+
+	/// Returns `true` if the list contains `id`, otherwise returns `false`.
+	pub(crate) fn contains(id: &T::AccountId) -> bool {
+		crate::ListNodes::<T>::contains_key(id)
+	}
+
+	/// Iterate over all nodes in all bags in the list.
+	///
+	/// Full iteration can be expensive; it's recommended to limit the number of items with
+	/// `.take(n)`.
+	pub(crate) fn iter() -> impl Iterator<Item = Node<T>> {
+		// We need a touch of special handling here: because we permit `T::BagThresholds` to
+		// omit the final bound, we need to ensure that we explicitly include that threshold in the
+		// list.
+		//
+		// It's important to retain the ability to omit the final bound because it makes tests much
+		// easier; they can just configure `type BagThresholds = ()`.
+		let thresholds = T::BagThresholds::get();
+		let iter = thresholds.iter().copied();
```

Comment:
```suggestion
		let iter = thresholds.into_iter();
```

doesn't this work?

Reply:
Threshold is `&[VoteWeight]` `into_iter` yields `&VoteWeight` because the slice cannot be dropped. So we still need to do `.copied()` to get derefed `VoteWeight`.

https://stackoverflow.com/questions/34733811/what-is-the-difference-between-iter-and-into-iter.